code/datums/components/hattable.dm

```diff
@@ -0,0 +1,110 @@
+TYPEINFO(/datum/component/hattable) // Take a walk through my TWISTED mind.... I'm sorry
+	initialization_args = list(
+		ARG_INFO("death_remove", DATA_INPUT_TEXT, "Remove component and signals on death?", FALSE), // Use this for things that don't revive, or explode on death
+		ARG_INFO("free_hat", DATA_INPUT_TEXT, "Should the parent get a free hat?", FALSE), // For AIs
+	)
+
+/datum/component/hattable  // Compatible stuff should be given hat_offset x and y vars in their own file
+	dupe_mode = COMPONENT_DUPE_HIGHLANDER
+	var/y_level = 0
+	var/x_level = 0
+	var/obj/item/hat = null
+	var/death_throw = FALSE
+	var/death_remove = FALSE
+	var/free_hat = FALSE
+	var/free_hat_lockout = FALSE
+
+/datum/component/hattable/Initialize(death_remove, free_hat)
+	. = ..()
+	RegisterSignal(src.parent, COMSIG_ATTACKBY, PROC_REF(hat_on_obj), override = TRUE)
+	src.death_throw = FALSE
+	src.death_remove = death_remove
+	src.free_hat = free_hat
+	if (free_hat && !free_hat_lockout) // If the thing gets a free hat (currently just AIs with spacebux) and they haven't received it yet, do that
+		var/free_hat_type = pick(childrentypesof(pick(childrentypesof(/obj/item/clothing/head))))
+		hat_on_obj(src.parent, free_hat_type)
+
+
+/datum/component/hattable/proc/hat_on_obj(mob/target as mob, obj/item/item as obj, mob/attacker)
+	var/obj/item/H = null
+	if (src.free_hat && !src.free_hat_lockout)
```

Comment:
Better to just make an instance in `Initialize` and pass it in here? You shouldn't need the lockout var then.